app/assets/javascripts/pages/bulkTable.js

```diff
@@ -280,15 +292,45 @@
     );
   };
 
+  // Entity -> State
+  state.addEntity = function(entity, orderIdx){
+    return state
+      .setIn(['entities', 'byId', entity.id], entity) // store entity
+      .spliceIntoOrder(entity.id, orderIdx) // order entity
+      .setIn(['matches', 'byEntityId', entity.id], {}) // create blank matches repo for entity
+      .setIn(['errors', 'byEntityId', entity.id], {}); // create blank errors repo
+  };
+
+  // String, Integer -> State
+  state.spliceIntoOrder = function(id, idx){
+    var order = state.getIn(['entities', 'order']);
+    return state.setIn(
+      ['entities', 'order'],
+      order
+        .slice(0, idx)
```

Comment:
i think you could use[ Array.splice()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice) here to make it a little shorter:

``` js
state.setIn(['entities', 'order'], state.getIn(['entities', 'order']).slice(idx, 0, id) )
```

Reply:
i considered that, but `#splice` mutates the array it is called on and then returns what it deleted, so you would always wind up with an array of the thing you were trying to remove!

Reply:
oh! i didn't realize that..jeez.

Reply:
yeah. way to go javascript. :/

the most terse solution for this operation i've come across involves using the spread operator:

```
[...arr.slice(0,n), newItem, ...arr.slice(n+1)] 
```

which is still not great.